### PR TITLE
gh-120478: `itertools.ilen` addition

### PR DESCRIPTION
**Pull request by dg-pb**
How initial implementation with tests would look like.

* Issue: gh-120478

### CODE REVIEW

File: Modules/itertoolsmodule.c [hunk 1]
```diff
@@ -3802,6 +3802,75 @@ static PyType_Spec ziplongest_spec = {
...
+    PyObject *new_long_cnt, *temp;
+    Py_ssize_t cnt;
+
+    it = PyObject_GetIter(iterable);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    cnt = 0;
+    iternext = *Py_TYPE(it)->tp_iternext;
```
Comment (picnixz) [inline Modules/itertoolsmodule.c:3832]: It might be helpful to possibly use the `__length_hint__` if available. For instance, if the latter gives you something way below the SSIZE_T_MAX, you can do a while loop without the `if` check and then another while loop with the `if`. More precisely, you can do a fast while loop (a while loop without `if cnt == PY_SSIZE_T_MAX`) up to `min(__length_hint__, PY_SSIZE_T_MAX)`, check if there are remaining items and if so, do the slow while loop.

~NB: In the slow while loop, you can split it into two as well where `long_cnt == NULL` and when it's been set once but since this condition only holds for a single `cnt` it does not really matter.~

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: Thanks for some food for thought.

The issue is... How would the fast loop look like?
```C
for (i=0; i<= PY_SSIZE_T_MAX, i++)
```
does the same `if` check implicitly.

Also, splitting into 2 loops where `long_cnt == NULL` and the "rest" will not result in much savings as the body of `if (cnt == PY_SSIZE_T_MAX) {` will be accessed once in 2147483647 on 32bit machine.

Also, comparison of timing if I replace PY_SSIZE_T_MAX with "1" so that it is accessed on every loop:
```python
a = range(100_000)
%timeit ilen(a)          # 1.45 ms
%timeit ilen_pylong(a)   # 3.72 ms
```

Comment (nineteendo) [reply Modules/itertoolsmodule.c:3832]: Hmm, wouldn't you need a for loop then?

```c
for (cnt = 0; cnt < length_hint && (item = iternext(it)) != NULL; cnt++) {
    Py_DECREF(item);
}
```

Is that even faster?

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: Checked it. Same runtime, no improvement at all.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]: > The issue is... How would the fast loop look like?

You can do a `while` and decrement `length_hint` in the loop condition itself and check for whether its `>0` or not. If at the end of the loop you hit `item != NULL` then you know that you've consumed `length_hint` items. If not, then you know that `length_hint` was an underestimation so you'll need to consume more, e.g.:

```C
rem = length_hint;
while ((item = iternext(it)) != NULL && --rem) {
	Py_DECREF(item);
}
// now either item == NULL or rem == 0
if (item != NULL) {
	// still things to process, do the slow loop
} else {
	cnt = length_hint - rem;
}
```

> Also, splitting into 2 loops where long_cnt == NULL and the "rest" will not result in much savings as the body of if (cnt == PY_SSIZE_T_MAX) { will be accessed once in 2147483647 on 32bit machine.

Yes, that's why I said it does not matter (I think I should not have mentioned at all :laughing:)

> Is that even faster?

This is something I am not sure about, hence the suggestion. I did not check the assembly myself for this one but I think it could be faster.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]: Ah I took too long for my answer. So using `__length_hint__` does not help at all, well sometimes I get surprises (now that I think about, we still have two ifs so it's maybe not really a surprise).

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: Hm, your code would be faster for sure, but there is an issue - what if `length_hint` is lying by a lot? There is a risk of overflow on a negative side.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]: - If it's underestimating, you'll reach the length_hint = 0 and exit the while loop so all's fine.
- If it's overestimating, you'll reach the `item == NULL` before and exit the while before being able to make `--length_hint` so you'll stay `>= 0`.

Note that you need to replace `length_hint` by `min(length_hint, PY_SSIZE_T_MAX)` so that you know that you'll not overflow on a negative by a lot. By the way, I did not checked my algorithm that I wrote (like I might have an 1-offset).

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: > * If it's underestimating, you'll reach the length_hint = 0 and exit the while loop so all's fine.

But there is no condition to exit at `0`, if we add one, then we loose performance gain.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]:
```C
while ((item = iternext(it)) != NULL && --rem) {
	Py_DECREF(item);
}
```

The condition is ``--rem``.

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: Then maybe no need for `length_hint` at all. Can just do `PY_SSIZE_T_MAX`.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]: Mmh, maybe yes. Though you'll need to handle the remainder differently then. But again, reversed loops are not always faster than forward loops, but if we strive for the fastest possible code (especially motivated by the purpose of `ilen`), then we should consider different alternatives. In particular, if the gain is minimal, then we can just stick to your first implementation.

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3832]: I can not observe any noticeable improvement. It still needs to increment and do a conditional check. If compiler is clever theoretically there shouldn't be any difference. Either way, I think gain here is too small to sacrifice simplicity.

Comment (picnixz) [reply Modules/itertoolsmodule.c:3832]: You can mark this discussion as resolved (sorry for a false hope).

File: Misc/NEWS.d/next/Library/2024-06-14-20-39-53.gh-issue-120478.q6lkDv.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Add :func:`ilen` to :mod:`itertools` module.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-06-14-20-39-53.gh-issue-120478.q6lkDv.rst]:
```suggestion
Add :func:`~itertools.ilen` to :mod:`itertools` module.
```

File: Doc/library/itertools.rst [hunk 3]
```diff
@@ -464,6 +465,23 @@ loops that truncate the stream.
...
+   Returns ``0`` for exhausted iterator.
+
+   Equivalent to::
+
+      def ilen(iterable):
+         length = 0
+         for _ in iterable:
+             length += 1
+         return length
+
```
Comment (picnixz) [inline Doc/library/itertools.rst]:
```suggestion
      def ilen(iterable):
         return len(list(iterable))
```

File: Doc/library/itertools.rst [hunk 4]
```diff
@@ -464,6 +465,23 @@ loops that truncate the stream.
                       pass
 
 
+.. function:: ilen(iterable)
+
+   Consume the iterator and return the number of consumed elements.
```
Comment (picnixz) [inline Doc/library/itertools.rst]: You are talking about an *iterator* but your argument is an *iterable*. I would phrase it like this:

```
Return the length of an *iterable*.

If *iterable* is an iterator, it is consumed.
```

File: Modules/itertoolsmodule.c [hunk 5]
```diff
@@ -3802,6 +3802,75 @@ static PyType_Spec ziplongest_spec = {
...
+itertools_ilen(PyObject *module, PyObject *iterable)
+/*[clinic end generated code: output=99a1c852a6d09dd4 input=cf51dec933b693da]*/
+{
+    PyObject *it, *item;
+    PyObject *(*iternext)(PyObject *);
+    PyObject *long_cnt = NULL;
+    PyObject *new_long_cnt, *temp;
+    Py_ssize_t cnt;
+
+    it = PyObject_GetIter(iterable);
```
Comment (picnixz) [inline Modules/itertoolsmodule.c:3826]: I'm just wondering, but is it possible to know whether an object is an iterable with known size or not in advance? for instance, if you pass `ilen([1,2,3])`, it should be equivalent to directly call `len` instead of having to count (again) the elements, right?

Comment (dg-pb) [reply Modules/itertoolsmodule.c:3826]: It is, but I don't think it is a good idea.

It is iterator module. If one has a list, he should call `len`.

E.g. `islice` function could jump straight to the index if a list is provided, but it doesn't. Because it is not what it is supposed to do.

It is good to be able to count on it that it will always call `iter` and iterate.

File: Misc/NEWS.d/next/Library/2024-06-14-20-39-53.gh-issue-120478.q6lkDv.rst (hunk 2 above)
Comment (dg-pb) [reply Misc/NEWS.d/next/Library/2024-06-14-20-39-53.gh-issue-120478.q6lkDv.rst]: Thank you!

File: Modules/itertoolsmodule.c (hunk 5 above)
Comment (picnixz) [reply Modules/itertoolsmodule.c:3826]: So maybe it should be good to say so? like add a recommendation about it (in the docs)

File: Doc/library/itertools.rst (hunk 3 above)
Comment (dg-pb) [reply Doc/library/itertools.rst]: `len(list())` is not a good equivalence. One of the main advantages of iterators is that one does not need to hold items in memory. Loop represents implementation much closer and speed does not matter here.

E.g. `islice` rough equivalence. It does not give `iter(list(iterable)[start:stop:sep])`, because it does not represent implementation, which is iteration oriented.

Comment (picnixz) [reply Doc/library/itertools.rst]: Then, `sum(1 for _ in iterable)`.

Comment (dg-pb) [reply Doc/library/itertools.rst]: Could maybe do:
```python
for n, _ in enumerate(iterable, 1):
    pass
return n
```

Also thought about
```python
sum(1 for _ in iterable)
```

But so far I like most basic loop best.

File: Modules/itertoolsmodule.c (hunk 5 above)
Comment (dg-pb) [reply Modules/itertoolsmodule.c:3826]: I have adjusted documentation to be more clear as per your suggestion above. Will push in a minute, let me know what you think. If it is ok or explicit recommendation to use `len` for sequences needed.

File: Doc/library/itertools.rst (hunk 3 above)
Comment (picnixz) [reply Doc/library/itertools.rst]: Honestly, I think it's a good practice to also highlight constructions that can be simplified. If you look at the recipes, that's what they do.

Comment (dg-pb) [reply Doc/library/itertools.rst]: But recipes are different, that is the whole point of them to make clever use of existing functions.

While "roughly equivalent" are here to represent implementation detail as accurately as possible.

I don't like `sum` because it calls `sum` function, which is not part of implementation.
I also don't like loop too much, because it has `_` in it, which doesn't look good.

File: Doc/library/itertools.rst [hunk 6]
```diff
@@ -464,6 +465,25 @@ loops that truncate the stream.
                       pass
 
 
+.. function:: ilen(iterable)
+
+   Return the length of an *iterable* by counting iterations.
```
Comment (picnixz) [inline Doc/library/itertools.rst]: I think it is obvious that you will count iterations to count the number of elements of an iterable (though, if you say "iterations" people might wonder "iterations of what"). So "Return the length of an *iterable*" should be sufficient.

File: Doc/library/itertools.rst (hunk 3 above)
Comment (picnixz) [reply Doc/library/itertools.rst]: Well.. I think we should wait for another opinion on that one.

File: Doc/library/itertools.rst [hunk 7]
```diff
@@ -464,6 +465,25 @@ loops that truncate the stream.
...
+
+   If *iterable* is an iterator, it is consumed.
```
Comment (picnixz) [inline Doc/library/itertools.rst]: Maybe the mention to `len` is not needed then. If `islice` does not make the distinction, then we should not mention it.

File: Doc/library/itertools.rst (hunk 6 above)
Comment (dg-pb) [reply Doc/library/itertools.rst]: What about "Iterate over *iterable* and return iteration count."?

Comment (picnixz) [reply Doc/library/itertools.rst]: I still think it's better to just explain what it does in simpler terms rather in a convoluted way here. "Return the length of the iterable" is straightforward and we explain just after that an iterator is consumed.

File: Doc/library/itertools.rst (hunk 3 above)
Comment (dg-pb) [reply Doc/library/itertools.rst]: I will go with `sum`. It does represent accurately enough and has a flavour of "it is as simple as". Loop is a bit of an overkill in verbosity here.

File: Doc/library/itertools.rst (hunk 6 above)
Comment (dg-pb) [reply Doc/library/itertools.rst]: But if the input is a `list`, then the user might be confused why it is taking so long, this will give a clearer signal that it achieves it by iterating over it. 

This would also address your previous concern better, to notify the user that it is not as efficient as `len` for `Sequences` and will provide an answer if one comes here to investigate.

Comment (picnixz) [reply Doc/library/itertools.rst]: Then you should mention this in a separate line, not in the description of the function itself.

File: Doc/library/itertools.rst (hunk 3 above)
Comment (nineteendo) [reply Doc/library/itertools.rst]: I think the for loop was fine. This looks like an optimisation, but isn't even one:

```shell
$ python -m timeit "length = 0" "for _ in range(100): length += 1"  
200000 loops, best of 5: 1.56 usec per loop
$ python -m timeit "length = sum(1 for _ in range(100))"  
100000 loops, best of 5: 3.34 usec per loop
```

The only other one-liner in the documentation is this:

```python
def compress(data, selectors):
    return (datum for datum, selector in zip(data, selectors) if selector)
```

And I would argue that this is more readable:

```python
def compress(data, selectors):
    for datum, selector in zip(data, selectors):
        if selector:
            yield datum
```

Comment (dg-pb) [reply Doc/library/itertools.rst]: I agree with you, `generator function` is much more friendly to my eyes.

If `try-exept` wasn't so cumbersome I would go with `while-try-except`, but it looks scary.

Switching back to `for` loop. It is the simplest after all.
